contrib/dockerd-rootless.sh

```diff
@@ -54,6 +54,16 @@ if ! [ -d "$HOME" ]; then
 	exit 1
 fi
 
+mount_directory() {
+	local directory="$1"
+	local mount_options="${2:---bind}"
+
+	directory_realpath=$(realpath "$directory")
+	rm -rf "$directory"
```

Comment:
It should be validated that `$directory` is inside the `copy-up` list?

Reply:
Commented on https://github.com/moby/moby/pull/48541#pullrequestreview-2341840519, we are not doing a `copy-up` as far as I understand. I extracted this logic to do a mount bind, but not necessarily a `copy-up` (well, `/etc` will be, but that's not a change introduced by this PR).

Please, let me know if I'm missing anything!

Reply:
`/etc` is copied up: https://github.com/moby/moby/blob/4001d0704ba38a82e1dbc26f0593fca66db1cb98/contrib/dockerd-rootless.sh#L136

`rm -rf $directory` w/o check might be safe with the current revision of the code, but can be potentially dangerous in future modifications to the script

Reply:
> but can be potentially dangerous in future modifications to the script

I am following the script with debugging and checking the PID and their mount namespace at all times, and when we delete the directories and remount them we are doing it always inside the pseudo-root namespace.

For example, `/var/run/cdi` wouldn't be on the copy-up list, but we still want to remount it inside the pseudo-root namespace. We could add it to the `copy-up` list, but we don't need to write to it, just read. Same in principle would go for `/etc/cdi`, but since we are copy-upping /etc it's fine I think, no need for more hassle.

Reply:
Maybe I could add some guardrails to `mount_directory` so that it's not called from the original namespace, even if the script evolves. I can look into what checks might give us that hint.

Reply:
If adding a guardrail is difficult, at least a comment line should be added

Reply:
Thanks @AkihiroSuda, let me give it some thought and I will ping you folks back on this PR.

Reply:
It was fairly trivial. I added the guardrails and now the script fails and exit if `mount_directory` is called from the wrong context.